src/grid/grid.component.ts

```diff
@@ -0,0 +1,630 @@
+
+    // child references
+    @ContentChildren(IgxColumnComponent) protected cols: QueryList<IgxColumnComponent>;
+    @ViewChild(IgxDialog) public editingModal: IgxDialog;
+    @ViewChild(IgxPaginatorComponent) public paginator: IgxPaginatorComponent;
+
+    // event emitters
+    @Output() public onEditDone = new EventEmitter<IgxGridEditEvent>();
+    @Output() public onFilterDone = new EventEmitter<IgxGridFilterEvent>();
+    @Output() public onSortingDone = new EventEmitter<IgxGridSortEvent>();
+    @Output() public onMovingDone = new EventEmitter();
+    @Output() public onCellSelection = new EventEmitter<IgxGridCellSelectionEvent>();
+    @Output() public onRowSelection = new EventEmitter<IgxGridRowSelectionEvent>();
+    @Output() public onPagingDone = new EventEmitter();
+    @Output() public onColumnInit = new EventEmitter<IgxGridColumnInitEvent>();
+    @Output() public onBeforeProcess = new EventEmitter();
+
+    get columnsToRender(): IgxColumnComponent[] {
+        return this.columns.filter((col) => !col.hidden);
+    }
+
+    get hasEditableColumns(): boolean {
+        return this.columns.some((col) => col.editable);
+    }
+
+    get hasSorting(): boolean {
+        return this.columns.some((col) => col.sortable);
+    }
+
+    get hasFiltering(): boolean {
+        return this.columns.some((col) => col.filtering);
+    }
+
+    get editableColumns(): IgxColumnComponent[] {
+        let res: IgxColumnComponent[] = [];
+        this.columns.forEach((col) => {
+            if (col.editable) {
+                res.push(col);
+            }
+        });
+        return res;
+    }
+    public dataContainer: DataContainer;
+    public columns: IgxColumnComponent[] = [];
+    protected selectedRow: any;
+    private filteringExpressions: Object = {};
+    private sortingExpressions: Object = {};
+    private colDiffer: Subscription;
+    private _differ: IterableDiffer;
+    private _paging: boolean = false;
+    private _lastRow: any;
+    private shouldDataBind: boolean = false;
+
+    constructor(private _renderer: Renderer,
+                private _elementRef: ElementRef,
+                private _detector: ChangeDetectorRef,
+                private _itDiff: IterableDiffers,
+                private _resolver: ComponentFactoryResolver,
+                private _viewRef: ViewContainerRef) {
+                    this._differ = this._itDiff.find([]).create(null);
+                }
+
+    public ngOnInit(): void {
+        if (this.data) {
+            this.dataContainer = new DataContainer(this.data);
+            if (this.paging) {
+                this.dataContainer.state.paging = {
+                    index: 0,
+                    recordsPerPage: this.rowsPerPage,
+                };
+            }
+            // should we attach to all browsers and only when we have sorting??
+            if (this.hasSorting) {
+                this.dataContainer.state.sorting.strategy = new StableSortingStrategy();
```

Comment:
Do a feature check. Perform a small sort to see if the browser already performs a stable sort. If it doesn't then attach our stable sorting strategy.